src/core/DY.Core.Bytes.fst

```diff
@@ -334,55 +362,78 @@ let get_dh_usage #cusages pk =
 /// by honest principals.
 
 noeq
-type crypto_predicates (cusages:crypto_usages) = {
-  aead_pred: tr:trace -> key:bytes{AeadKey? (get_usage key)} -> nonce:bytes -> msg:bytes -> ad:bytes -> prop;
-  aead_pred_later:
+type aead_crypto_predicate (cusages:crypto_usages) = {
+  pred: tr:trace -> key:bytes{AeadKey? (get_usage key)} -> msg:bytes -> ad:bytes -> prop;
+  pred_later:
     tr1:trace -> tr2:trace ->
-    key:bytes{AeadKey? (get_usage key)} -> nonce:bytes -> msg:bytes -> ad:bytes ->
+    key:bytes{AeadKey? (get_usage key)} -> msg:bytes -> ad:bytes ->
     Lemma
-    (requires aead_pred tr1 key nonce msg ad /\ tr1 <$ tr2)
-    (ensures aead_pred tr2 key nonce msg ad)
+    (requires pred tr1 key msg ad /\ tr1 <$ tr2)
+    (ensures pred tr2 key msg ad)
   ;
+}
 
-  pkenc_pred: tr:trace -> pk:bytes{PkdecKey? (get_sk_usage pk)} -> msg:bytes -> prop;
-  pkenc_pred_later:
+noeq
+type pkenc_crypto_predicate (cusages:crypto_usages) = {
+  pred: tr:trace -> pk:bytes{PkdecKey? (get_sk_usage pk)} -> msg:bytes -> prop;
+  pred_later:
     tr1:trace -> tr2:trace ->
     pk:bytes{PkdecKey? (get_sk_usage pk)} -> msg:bytes ->
     Lemma
-    (requires pkenc_pred tr1 pk msg /\ tr1 <$ tr2)
-    (ensures pkenc_pred tr2 pk msg)
+    (requires pred tr1 pk msg /\ tr1 <$ tr2)
+    (ensures pred tr2 pk msg)
   ;
+}
 
-  sign_pred: tr:trace -> vk:bytes{SigKey? (get_signkey_usage vk)} -> msg:bytes -> prop;
-  sign_pred_later:
+noeq
+type sign_crypto_predicate (cusages:crypto_usages) = {
+  pred: tr:trace -> vk:bytes{SigKey? (get_signkey_usage vk)} -> msg:bytes -> prop;
+  pred_later:
     tr1:trace -> tr2:trace ->
     vk:bytes{SigKey? (get_signkey_usage vk)} -> msg:bytes ->
     Lemma
-    (requires sign_pred tr1 vk msg /\ tr1 <$ tr2)
-    (ensures sign_pred tr2 vk msg)
+    (requires pred tr1 vk msg /\ tr1 <$ tr2)
+    (ensures pred tr2 vk msg)
   ;
+}
 
-  // ...
+noeq
+type crypto_predicates (cusages:crypto_usages) = {
+  aead_pred: aead_crypto_predicate cusages;
+  pkenc_pred: pkenc_crypto_predicate cusages;
+  sign_pred: sign_crypto_predicate cusages;
 }
 
 /// Default (empty) cryptographic predicates, that can be used like this:
 /// { default_crypto_predicates with
 ///   sign_pred = ...;
-///   sign_pred_later = ...;
 /// }
 
+val default_aead_predicate: cusages:crypto_usages -> aead_crypto_predicate cusages
+let default_aead_predicate cusages = {
+  pred = (fun tr key msg ad -> False);
```

Comment:
should the default be `False` or `True` ?

Reply:
I guess it doesn't change much. Currently, all default predicates in DY* are `False`, so for consistency I also used `False` here.
- `False` means that honest principals are not allowed to compute any cryptographic function, and the predicates allows them in some particular cases.
- `True` means that honest principal are allowed to compute any cryptographic function, and the predicates restrict them in some particular case.

If we prefer the second option it would be fit another pull request, but unless there is a good technical reason I don't think it's worth spending time on that

Reply:
In my opinion, keeping `False` makes more sense.

Reply:
I think, conceptually `False` is maybe the better option (be as restrictive as possible and only allow what is needed).
On the other hand, when I started to learn DY*, I first set all predicates to `True` so that I don't have to worry, whether I'm allowed to call the functions or not. And only later, when trying to show the invariant, I restricted the predicates. But maybe this was also part of the intrinsic version and is now resolved with the extrinsic one.
I just wanted to bring it up and see opinions. Using `False` consistently is fine for me.